README.md
# Flutter Web Custom HTML Element Widget

A widget that allows you to use custom html elements tracked on top of your flutter widget.

> This package is Web only!

## Features

Create a custom html element through javascript and track it on top of your flutter widget. 
You can change size, position or make it part of scrollable views and it will follow the widget.
There is a bit of a delay, but we are looking for a way to fix that.

You can also change the attributes, which will change the html elements attributes.

## Getting started


Add the following to your pubspec.yaml
```yaml
custom_html_element_web:
  git:
    url: git@bitbucket.org:iconicadevs/flutter_custom_html_element_web.git
    ref: 0.0.1
```

## Usage

See the example for more Information.

```dart
return CustomHtmlElement(
    width: 160,
    height: 90.0 + increment,
    tag: 'my-counter',
    // by providing the controller, the html element can update
    // every time the scroll controller updates.
    layoutObservable: controller,
    attributes: {
        'data-initial-count': '12',
        'data-increment-amount': '$increment',
        'data-on-count-changed': 'onUpdateCount',
    },
);
```

The provided tag needs to be an html element that already exists or that you created. 
The example also shows how to communicate between your custom element and your 

## Issues

Please file any issues, bugs or feature request as an issue on our [GitHub](https://github.com/Iconica-Development/flutter_community_chat/pulls) page. Commercial support is available if you need help with integration with your app or services. You can contact us at [support@iconica.nl](mailto:support@iconica.nl).

## Want to contribute

If you would like to contribute to the plugin (e.g. by improving the documentation, solving a bug or adding a cool new feature), please carefully review our [contribution guide](../CONTRIBUTING.md) and send us your [pull request](https://github.com/Iconica-Development/flutter_community_chat/pulls).

## Author

This `custom_html_element_web` for Flutter is developed by [Iconica](https://iconica.nl). You can contact us at <support@iconica.nl>

CHANGELOG.md
## 0.0.1

Add Basic custom HTML element functionality:

- Add scroll support
- Add attribute watching
- Add size and offset watching


lib/src/custom_html_element.dart
// ignore: avoid_web_libraries_in_flutter
import 'dart:html' as html;

import 'package:flutter/material.dart';

/// A widget that allows you to create a custom HTML element
/// that follows this widget position and size.
///
/// This is meant for custom HTML elements that require custom attributes
/// and cannot be rendered through other flutter html packages.
///
/// It is not possible to provide children to the custom HTML element, for now.
class CustomHtmlElement extends StatefulWidget {
  const CustomHtmlElement({
    required this.tag,
    required this.height,
    required this.width,
    this.layoutObservable,
    this.attributes = const {},
    super.key,
  });

  /// the width of the widget
  final double width;

  /// the height of the widget;
  final double height;

  /// The tag name of the html element
  ///
  /// for example, a <p></p> will be a p.
  final String tag;
  final Map<String, String> attributes;

  /// A listenable that notifies the widget that its position has changed.
  ///
  /// For example, providing a scroll controller here will cause the related
  /// HTML element to be moved to the correct spot according to the flutter
  /// location of the widget.
  final Listenable? layoutObservable;

  @override
  State<CustomHtmlElement> createState() => _CustomHtmlElementState();
}

class _CustomHtmlElementState extends State<CustomHtmlElement>
    with WidgetsBindingObserver {
  late final html.Element _internalElement;
  GlobalKey _key = GlobalKey();

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addObserver(this);
    WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
      _addInternalElement();
    });

    widget.layoutObservable?.addListener(_onLayoutObservableChanged);
  }

  void _addInternalElement() {
    _internalElement = html.document.createElement(widget.tag);
    if (!mounted) {
      return;
    }

    _internalElement.style.position = 'fixed';
    _internalElement.style.display = 'block';
    _setOffset();

    for (var attribute in widget.attributes.entries) {
      _internalElement.setAttribute(attribute.key, attribute.value);
    }
    html.document.querySelector('body')?.append(_internalElement);
  }

  void _setOffset() {
    var renderBox = _key.currentContext?.findRenderObject() as RenderBox;

    final offset = renderBox.localToGlobal(Offset.zero);
    _internalElement.style.left = '${offset.dx}px';
    _internalElement.style.top = '${offset.dy}px';
    _internalElement.style.width = '${widget.width}px';
    _internalElement.style.height = '${widget.height}px';
  }

  DateTime _observableLastChanged = DateTime.now();

  void _onLayoutObservableChanged() {
    _setOffset();
    var lastChanged = _observableLastChanged = DateTime.now();
    // autoCorrect after 50 milliseconds;
    Future.delayed(const Duration(milliseconds: 50), () {
      if (_observableLastChanged == lastChanged && mounted) {
        _setOffset();
      }
    });
  }

  @override
  void dispose() {
    super.dispose();

    WidgetsBinding.instance.removeObserver(this);

    widget.layoutObservable?.removeListener(_onLayoutObservableChanged);
    _internalElement.remove();
  }

  @override
  void didChangeMetrics() {
    _setOffset();
  }

  @override
  void didUpdateWidget(covariant CustomHtmlElement oldWidget) {
    super.didUpdateWidget(oldWidget);

    if (!mounted) {
      return;
    }

    if (oldWidget.layoutObservable != widget.layoutObservable) {
      oldWidget.layoutObservable?.removeListener(_onLayoutObservableChanged);
      widget.layoutObservable?.addListener(_onLayoutObservableChanged);
    }

    if (oldWidget.width != widget.width || oldWidget.height != widget.height) {
      _setOffset();
      setState(() {
        _key = GlobalKey();
      });
    }

    var delta = oldWidget.attributes.getDelta(widget.attributes);
    for (var attribute in delta.entries) {
      var value = attribute.value;
      if (value == null) {
        _internalElement.removeAttribute(attribute.key);
      } else {
        _internalElement.setAttribute(attribute.key, value);
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      key: _key,
      width: widget.width,
      height: widget.height,
    );
  }
}

extension _MapDelta<K, V> on Map<K, V> {
  /// Get the delta between two maps
  ///
  /// A null value indicates that the value exists in this map,
  /// but that [other] does not contain that value
  ///
  /// This does not perform a recursive delta
  Map<K, V?> getDelta(Map<K, V> other) {
    var delta = <K, V?>{};
    for (var entry in entries) {
      if (!other.containsKey(entry.key)) {
        delta[entry.key] = null;
      }
      if (other[entry.key] != entry.value) {
        delta[entry.key] = other[entry.key];
      }
    }
    for (var entry in other.entries) {
      if (!containsKey(entry.key)) {
        delta[entry.key] = entry.value;
      }
    }
    return delta;
  }
}


example/lib/main.dart
import 'package:custom_html_element_web/custom_html_element.dart';
import 'package:flutter/material.dart';
import 'package:js/js.dart';

@JS('onUpdateCount')
external set _onUpdateCount(void Function(dynamic) function);

// this is called every time the counter increases
void onUpdateCounter(count) {
  debugPrint(count.toString());
}

void main() {
  _onUpdateCount = allowInterop(onUpdateCounter);
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({
    required this.title,
    super.key,
  });

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int increment = 1;

  final controller = ScrollController();

  @override
  void dispose() {
    super.dispose();
    controller.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: ListView.builder(
        controller: controller,
        itemBuilder: (context, value) {
          if (value == 10) {
            // we create only one HTML element for testing
            return CustomHtmlElement(
              width: 160,
              height: 90.0 + increment,
              tag: 'my-counter',
              // by providing the controller, the html element can update
              // every time the scroll controller updates.
              layoutObservable: controller,
              attributes: {
                'data-initial-count': '12',
                'data-increment-amount': '$increment',
                'data-on-count-changed': 'onUpdateCount',
              },
            );
          }
          if (value.isEven) {
            return Container(
              height: 100,
              width: 100,
              color: Colors.red,
            );
          }
          return Container(
            height: 100,
            width: 100,
            color: Colors.blue,
          );
        },
      ),
      floatingActionButton: FloatingActionButton.small(
        onPressed: () {
          setState(() {
            increment += 1;
          });
        },
      ),
    );
  }
}
